upgrade tools

hawkingrei opened the pull request:
<!-- Thanks for sending a PR! Before submitting:

1. If this is your first PR, please read CONTRIBUTING.md and sign the CLA
   first. We cannot review code without a signed CLA.
2. Please file an issue *first*. All features and most bug fixes should have
   an associated issue with a design discussed and decided upon. Small bug
   fixes and documentation improvements don't need issues.
3. New features and bug fixes must have tests. Documentation may need to
   be updated. If you're unsure what to update, send the PR, and we'll discuss
   in review.
4. Note that PRs updating dependencies and new Go versions are not accepted.
   Please file an issue instead.
-->

**What type of PR is this?**

> Uncomment one line below and remove others.
>
> Bug fix
> Feature
> Documentation
> Other

**What does this PR do? Why is it needed?**

```x/tools``` is too old to upgrade the linter such as golangci-linter. 

**Which issues(s) does this PR fix?**

Fixes #3640

**Other notes for review**

hawkingrei commented:
@sluongng I've decided to push this forward, it seems that it needs to have an upload to mirror?

sluongng commented:
Yeah use this to request for a mirror https://github.com/bazelbuild/bazel/issues/new?assignees=sgowroji%2Cpavank1992%2Ciancha1992&labels=mirror+request%2Ctype%3A+process%2CP2%2Cteam-OSS&projects=&template=mirror_request.yml&title=%5BMirror%5D+

hawkingrei commented on tests/integration/popular_repos/README.rst:
```diff
@@ -191,8 +191,6 @@ This runs tests from the repository `golang.org/x/tools <https://golang.org/x/to
 * @org_golang_x_tools//internal/jsonrpc2:jsonrpc2_test
 * @org_golang_x_tools//internal/jsonrpc2/servertest:servertest_test
 * @org_golang_x_tools//internal/jsonrpc2_v2:jsonrpc2_v2_test
-* @org_golang_x_tools//internal/lockedfile:lockedfile_test
```
I cannot find this package in the ```x/tools```.

hawkingrei commented:
@sluongng CI has passed.

hawkingrei commented:
@fmeum PTAL

sluongng reviewed (approved):
LGTM

Curious readers might want to give https://github.com/golang/tools/releases a read for latest features. A lot of them are `gopls` related, which are irrelevant to rules_go, but the analysis and memory improvements are interesting regardless.

tyler-french commented on go/tools/builders/nogo_main.go:
```diff
@@ -609,8 +609,8 @@ func (i *importer) Import(path string) (*types.Package, error) {
 
-func (i *importer) readFacts(pkg *types.Package) ([]byte, error) {
-	archive := i.factMap[pkg.Path()]
+func (i *importer) readFacts(kgPath string) ([]byte, error) {
```
What does `kgPath` mean semantically?

hawkingrei replied:
It should be ```pkgPath```. This is my fault.

tyler-french commented on go/tools/builders/nogo_main.go:
```diff
@@ -234,7 +234,7 @@ func checkPackage(analyzers []*analysis.Analyzer, packagePath string, packageFil
 	// Process diagnostics and encode facts for importers of this package.
 	diagnostics := checkAnalysisResults(roots, pkg)
-	facts := pkg.facts.Encode()
+	facts := pkg.facts.Encode(true)
```
```suggestion
	facts := pkg.facts.Encode(true /* skipMethodSorting */)
```
Could we either attach a meaning to this? I don't want future confusion from API breaking changes to this call.

Naked parameters in function calls can hurt readability.

hawkingrei replied:
Added

tyler-french commented on go/tools/builders/nogo_main.go:
```diff
@@ -396,7 +396,7 @@ func load(packagePath string, imp *importer, filenames []string) (*goPackage, er
 	pkg.types, pkg.typesInfo = types, info
 
-	pkg.facts, err = facts.NewDecoder(pkg.types).Decode(imp.readFacts)
+	pkg.facts, err = facts.NewDecoder(pkg.types).Decode(true, imp.readFacts)
```
```suggestion
	pkg.facts, err = facts.NewDecoder(pkg.types).Decode(true  /* skipMethodSorting */, imp.readFacts)
```

Similar to above. Naked parameters in function calls can hurt readability.

hawkingrei replied:
Added

tyler-french reviewed (commented):
> Thanks for taking this on! Looks good to me.
> 
> @tyler-french Could you give this a test run?

@fmeum working on testing this now. 

@hawkingrei Could you `git rebase -i` and squash this PR into one commit? This might make it a bit easier to test.

Also added a couple of comments.

hawkingrei commented:
> > Thanks for taking this on! Looks good to me.
> > @tyler-french Could you give this a test run?
> 
> @fmeum working on testing this now.
> 
> @hawkingrei Could you `git rebase -i` and squash this PR into one commit? This might make it a bit easier to test.
> 
> Also added a couple of comments.

I have rebased it.

fmeum commented:
@hawkingrei Thanks!

hawkingrei commented:
> @hawkingrei Thanks!

When will we release a new rules_go on the github release ?

fmeum commented:
@tyler-french is already working on it
